File: rust/katari-runtime/src/runtime/primitive.rs
use crate::value::Value;

/// Result of a primitive call.
pub enum PrimitiveResult {
    /// Synchronous result value.
    Ok(Value),
    /// Raise a request (e.g. prim.parse_error).
    RaiseRequest { req_name: String, args: Vec<Value> },
}

fn parse_error(msg: String) -> PrimitiveResult {
    PrimitiveResult::RaiseRequest {
        req_name: "prim.parse_error".to_string(),
        args: vec![Value::String(msg)],
    }
}

/// Execute a primitive agent synchronously.
pub fn call_primitive(name: &str, args: &[Value]) -> PrimitiveResult {
    let val = |v| PrimitiveResult::Ok(v);

    match name {
        "prim.to_string" => {
            let v = args.first().unwrap_or(&Value::Null);
            val(Value::String(v.to_display_string()))
        }
        "prim.div" => val(match (args.first(), args.get(1)) {
            (Some(Value::Integer(a)), Some(Value::Integer(b))) => {
                if *b == 0 {
                    Value::Null
                } else {
                    Value::Integer(a.div_euclid(*b))
                }
            }
            (Some(a), Some(b)) => match (a.to_f64(), b.to_f64()) {
                (Some(a), Some(b)) if b != 0.0 => Value::Integer((a / b).floor() as i64),
                _ => Value::Null,
            },
            _ => Value::Null,
        }),
        "prim.mod" => val(match (args.first(), args.get(1)) {
            (Some(Value::Integer(a)), Some(Value::Integer(b))) => {
                if *b == 0 {
                    Value::Null
                } else {
                    Value::Integer(a.rem_euclid(*b))
                }
            }
            (Some(a), Some(b)) => match (a.to_f64(), b.to_f64()) {
                (Some(a), Some(b)) if b != 0.0 => Value::Number(a % b),
                _ => Value::Null,
            },
            _ => Value::Null,
        }),
        "prim.parse_integer" => match args.first() {
            Some(Value::String(s)) => match s.trim().parse::<i64>() {
                Ok(n) => val(Value::Integer(n)),
                Err(_) => parse_error(format!("failed to parse '{}' as integer", s)),
            },
            _ => parse_error("parse_integer: expected string argument".to_string()),
        },
        "prim.parse_number" => match args.first() {
            Some(Value::String(s)) => match s.trim().parse::<f64>() {
                Ok(n) => val(Value::Number(n)),
                Err(_) => parse_error(format!("failed to parse '{}' as number", s)),
            },
            _ => parse_error("parse_number: expected string argument".to_string()),
        },
        "prim.parse_boolean" => match args.first() {
            Some(Value::String(s)) => match s.as_str() {
                "true" => val(Value::Boolean(true)),
                "false" => val(Value::Boolean(false)),
                _ => parse_error(format!("failed to parse '{}' as boolean", s)),
            },
            _ => parse_error("parse_boolean: expected string argument".to_string()),
        },
        "prim.log.info" => {
            if let Some(Value::String(msg)) = args.first() {
                tracing::info!("{}", msg);
            }
            val(Value::Null)
        }
        "prim.log.warn" => {
            if let Some(Value::String(msg)) = args.first() {
                tracing::warn!("{}", msg);
            }
            val(Value::Null)
        }
        "prim.log.error" => {
            if let Some(Value::String(msg)) = args.first() {
                tracing::error!("{}", msg);
            }
            val(Value::Null)
        }
        "prim.length" => val(match args.first() {
            Some(Value::Array(arr)) => Value::Integer(arr.len() as i64),
            _ => Value::Integer(0),
        }),
        "prim.slice" => val(
            match (args.first(), args.get(1), args.get(2)) {
                (Some(Value::Array(arr)), Some(Value::Integer(start)), Some(Value::Integer(end))) => {
                    let s = (*start).max(0) as usize;
                    let e = (*end).max(0) as usize;
                    let s = s.min(arr.len());
                    let e = e.min(arr.len());
                    if s <= e {
                        Value::Array(arr[s..e].to_vec())
                    } else {
                        Value::Array(vec![])
                    }
                }
                _ => Value::Array(vec![]),
            },
        ),
        _ => {
            tracing::warn!("unknown primitive agent: {}", name);
            val(Value::Null)
        }
    }
}

File: ts/packages/katari-runtime/src/primitive.ts
import type { Value } from "./value.js";
import { toDisplayString } from "./value.js";

export type PrimitiveResult =
  | { tag: "Ok"; value: Value }
  | { tag: "RaiseRequest"; reqName: string; args: Value[] };

function ok(value: Value): PrimitiveResult {
  return { tag: "Ok", value };
}

function parseError(msg: string): PrimitiveResult {
  return { tag: "RaiseRequest", reqName: "prim.parse_error", args: [msg] };
}

export function callPrimitive(name: string, args: Value[]): PrimitiveResult {
  const a0 = args[0] ?? null;
  const a1 = args[1] ?? null;
  const a2 = args[2] ?? null;

  switch (name) {
    case "prim.to_string":
      return ok(toDisplayString(a0));

    case "prim.div": {
      if (typeof a0 === "number" && typeof a1 === "number") {
        if (a1 === 0) return ok(null);
        if (Number.isInteger(a0) && Number.isInteger(a1)) {
          // Euclidean floor division for integers
          return ok(Math.floor(a0 / a1) + (a0 % a1 !== 0 && (a0 ^ a1) < 0 ? 0 : 0));
        }
        return ok(Math.trunc(Math.floor(a0 / a1)));
      }
      return ok(null);
    }

    case "prim.mod": {
      if (typeof a0 === "number" && typeof a1 === "number") {
        if (a1 === 0) return ok(null);
        return ok(((a0 % a1) + a1) % a1);
      }
      return ok(null);
    }

    case "prim.parse_integer":
      if (typeof a0 === "string") {
        const n = parseInt(a0.trim(), 10);
        return isNaN(n) ? parseError(`failed to parse '${a0}' as integer`) : ok(n);
      }
      return parseError("parse_integer: expected string argument");

    case "prim.parse_number":
      if (typeof a0 === "string") {
        const n = parseFloat(a0.trim());
        return isNaN(n) ? parseError(`failed to parse '${a0}' as number`) : ok(n);
      }
      return parseError("parse_number: expected string argument");

    case "prim.parse_boolean":
      if (typeof a0 === "string") {
        if (a0 === "true") return ok(true);
        if (a0 === "false") return ok(false);
        return parseError(`failed to parse '${a0}' as boolean`);
      }
      return parseError("parse_boolean: expected string argument");

    case "prim.log.info":
      if (typeof a0 === "string") console.log(`[INFO] ${a0}`);
      return ok(null);

    case "prim.log.warn":
      if (typeof a0 === "string") console.warn(`[WARN] ${a0}`);
      return ok(null);

    case "prim.log.error":
      if (typeof a0 === "string") console.error(`[ERROR] ${a0}`);
      return ok(null);

    case "prim.length":
      return ok(Array.isArray(a0) ? a0.length : 0);

    case "prim.slice": {
      if (Array.isArray(a0) && typeof a1 === "number" && typeof a2 === "number") {
        const s = Math.min(Math.max(0, a1), a0.length);
        const e = Math.min(Math.max(0, a2), a0.length);
        return ok(s <= e ? a0.slice(s, e) : []);
      }
      return ok([]);
    }

    default:
      console.warn(`unknown primitive agent: ${name}`);
      return ok(null);
  }
}
